src/jit/gentree.cpp

```diff
@@ -6814,105 +7071,112 @@ void Compiler::gtBlockOpInit(
+    result->gtFlags |= GTF_REVERSE_OPS;
 
-        // If the class being copied contains any GC pointer we store a class handle
-        // and we must set the flag GTF_BLK_HASGCPTR, so that the register predictor
-        // knows that this GT_COPYBLK will use calls to the ByRef Assign helper
-        //
-        if ((hndOrSize->OperGet() == GT_CNS_INT) && hndOrSize->IsIconHandle(GTF_ICON_CLASS_HDL))
-        {
-            hndOrSize->gtFlags |= GTF_DONT_CSE; // We can't CSE the class handle
-            result->gtFlags |= GTF_BLK_HASGCPTR;
-        }
-    }
+    result->gtFlags |= (dst->gtFlags & GTF_EXCEPT) | (srcOrFillVal->gtFlags & GTF_EXCEPT);
 
-    if (volatil)
+    if (isVolatile)
     {
         result->gtFlags |= GTF_BLK_VOLATILE;
     }
 
 #ifdef FEATURE_SIMD
-    if (oper == GT_COPYBLK && srcOrFillVal->OperGet() == GT_ADDR && dst->OperGet() == GT_ADDR)
+    if (result->OperIsCopyBlkOp() && varTypeIsSIMD(srcOrFillVal))
     {
         // If the source is a GT_SIMD node of SIMD type, then the dst lclvar struct
         // should be labeled as simd intrinsic related struct.
         // This is done so that the morpher can transform any field accesses into
         // intrinsics, thus avoiding conflicting access methods (fields vs. whole-register).
 
-        GenTreePtr srcChild = srcOrFillVal->gtGetOp1();
-        GenTreePtr dstChild = dst->gtGetOp1();
-
-        if (dstChild->OperIsLocal() && varTypeIsStruct(dstChild) && srcChild->OperGet() == GT_SIMD &&
-            varTypeIsSIMD(srcChild))
+        GenTree* src = srcOrFillVal;
+        if (src->OperIsIndir() && (src->AsIndir()->Addr()->OperGet() == GT_ADDR))
         {
-            unsigned   lclNum                = dst->gtGetOp1()->AsLclVarCommon()->GetLclNum();
-            LclVarDsc* lclVarDsc             = &lvaTable[lclNum];
-            lclVarDsc->lvUsedInSIMDIntrinsic = true;
+            src = src->AsIndir()->Addr()->gtGetOp1();
+        }
+        if (src->OperGet() == GT_SIMD)
+        {
+            if (dst->OperIsBlk() && (dst->AsIndir()->Addr()->OperGet() == GT_ADDR))
+            {
+                dst = dst->AsIndir()->Addr()->gtGetOp1();
+            }
+
+            if (dst->OperIsLocal() && varTypeIsStruct(dst))
+            {
+                unsigned   lclNum                = dst->AsLclVarCommon()->GetLclNum();
+                LclVarDsc* lclVarDsc             = &lvaTable[lclNum];
+                lclVarDsc->lvUsedInSIMDIntrinsic = true;
+            }
         }
     }
 #endif // FEATURE_SIMD
 }
 
 //------------------------------------------------------------------------
-// gtNewBlkOpNode: Creates an InitBlk or CpBlk node.
+// gtNewBlkOpNode: Creates a GenTree for a block (struct) assignment.
 //
 // Arguments:
-//    oper          - GT_COPYBLK, GT_INITBLK or GT_COPYOBJ
 //    dst           - Destination or target to copy to / initialize the buffer.
```

Comment:
If I understood correctly, here dst has to be either a GT_OBJ/BLK/DYN_BLK or a LclVar?

If yes, can we be more clear in function header comments and also assert the same in this routine?